doc/gsl-randist.1
.\" Man page contributed by Dirk Eddelbuettel <edd@debian.org>
.\" and released under the GNU General Public License
.TH GSL-RANDIST 1 "" GNU
.SH NAME
gsl-randist - generate random samples from various distributions
.SH SYNOPSYS
.B gsl-randist seed n DIST param1 param2 [..]
.SH DESCRIPTION
.B gsl-randist 
is a demonstration program for the GNU Scientific Library.
It generates n random samples from the distribution DIST using the distribution
parameters param1, param2, ...
.SH EXAMPLE
Here is an example.  We generate 10000 random samples from a Cauchy
distribution with a width of 30 and histogram them over the range -100 to
100, using 200 bins.
 
     gsl-randist 0 10000 cauchy 30 | gsl-histogram -100 100 200 > histogram.dat
 
A plot of the resulting histogram will show the familiar shape of the
Cauchy distribution with fluctuations caused by the finite sample
size.

     awk '{print $1, $3 ; print $2, $3}' histogram.dat | graph -T X

.SH SEE ALSO
.BR gsl(3) ,
.BR gsl-histogram(1) .

.SH AUTHOR
.B gsl-randist 
was written by James Theiler and Brian Gough.
Copyright 1996-2000; for copying conditions see the GNU General
Public Licence. 

This manual page was added by the Dirk Eddelbuettel
<edd@debian.org>, the Debian GNU/Linux maintainer.


NEWS
* What is new in gsl-2.9:

* What was new in gsl-2.8:

** apply patch for bug #63679 (F. Weimer)

** updated multilarge TSQR method to store ||z_2|| and
   provide it to the user

** add routines for Hermite B-spline interpolation

** fix for bug #59624

** fix for bug #59781 (M. Dunlap)

** bug fix #61094 (reported by A. Cheylus)

** add functions:
   - gsl_matrix_complex_conjugate
   - gsl_vector_complex_conj_memcpy
   - gsl_vector_complex_div_real
   - gsl_linalg_QR_lssolvem_r
   - gsl_linalg_complex_QR_lssolvem_r
   - gsl_linalg_complex_QR_QHmat_r
   - gsl_linalg_QR_UR_lssolve
   - gsl_linalg_QR_UR_lssvx
   - gsl_linalg_QR_UR_QTvec
   - gsl_linalg_QR_UU_lssvx
   - gsl_linalg_QR_UD_lssvx
   - gsl_linalg_QR_UD_QTvec
   - gsl_linalg_complex_cholesky_{decomp2,svx2,solve2,scale,scale_apply}
   - gsl_linalg_SV_{solve2,lssolve}
   - gsl_rstat_norm

** add Lebedev quadrature (gsl_integration_lebedev)

** major overhaul to the B-spline module to add
   new functionality

* What was new in gsl-2.7.1:

** update libtool version numbers

* What was new in gsl-2.7:

** fixed doc bug for gsl_histogram_min_bin (lhcsky at 163.com)

** fixed bug #60335 (spmatrix test failure, J. Lamb)

** fixed bug #36577

** clarified documentation on interpolation accelerators (V. Krishnan)

** fixed bug #45521 (erroneous GSL_ERROR_NULL in ode-initval2, thanks to M. Sitte)

** fixed doc bug #59758

** fixed bug #58202 (rstat median for n=5)

** added support for native C complex number types in gsl_complex
   when using a C11 compiler

** upgraded to autoconf 2.71, automake 1.16.3, libtool 2.4.6

** updated exponential fitting example for nonlinear least squares

** added banded LU decomposition and solver (gsl_linalg_LU_band)

** New functions added to the library:
      - gsl_matrix_norm1
      - gsl_spmatrix_norm1
      - gsl_matrix_complex_conjtrans_memcpy
      - gsl_linalg_QL: decomp, unpack
      - gsl_linalg_complex_QR_* (thanks to Christian Krueger)
      - gsl_vector_sum
      - gsl_matrix_scale_rows
      - gsl_matrix_scale_columns
      - gsl_multilarge_linear_matrix_ptr
      - gsl_multilarge_linear_rhs_ptr
      - gsl_spmatrix_dense_add (renamed from gsl_spmatrix_add_to_dense)
      - gsl_spmatrix_dense_sub
      - gsl_linalg_cholesky_band: solvem, svxm, scale, scale_apply
      - gsl_linalg_QR_UD: decomp, lssolve
      - gsl_linalg_QR_UU: decomp, lssolve, QTvec
      - gsl_linalg_QR_UZ: decomp
      - gsl_multifit_linear_lcurvature
      - gsl_spline2d_eval_extrap

** bug fix in checking vector lengths in gsl_vector_memcpy (dieggsy@pm.me)

** made gsl_sf_legendre_array_index() inline and documented
   gsl_sf_legendre_nlm()

* What was new in gsl-2.6:

** add BLAS calls for the following functions:
     - gsl_vector_memcpy
     - gsl_vector_scale
     - gsl_matrix_memcpy
     - gsl_matrix_transpose_memcpy
     - gsl_matrix_tricpy
     - gsl_matrix_transpose_tricpy

** deprecated functions gsl_linalg_complex_householder_hm and
   gsl_linalg_complex_householder_mh

** add unit tests for gsl_linalg_symmtd and gsl_linalg_hermtd

** multilarge TSQR algorithm has been converted to use the new Level 3 QR decomposition

** nonlinear least squares Cholesky solver now uses the new Level 3 BLAS
   method; the old modified Cholesky solver is still available under
   gsl_multifit_nlinear_solver_mcholesky and gsl_multilarge_nlinear_solver_mcholesky

** implemented Level 3 BLAS versions of several linear algebra routines:
     - Triangular matrix inversion
     - Cholesky decomposition and inversion (real and complex)
     - LU decomposition and inversion (real and complex)
     - QR decomposition (courtesy of Julien Langou)
     - Generalized symmetric/hermitian eigensystem reduction to standard form

** removed deprecated function gsl_linalg_hessenberg()

** renamed gsl_interp2d_eval_e_extrap() to gsl_interp2d_eval_extrap_e()
   to match documentation (reported by D. Lebrun-Grandie)

** renamed some of the gsl_sf_hermite functions to be more consistent
   with rest of the library, and deprecated old function names

** updated gsl_sf_hermite_func() to use a newer algorithm
   due to B. Bunck which is more stable for large x; also added
   gsl_sf_hermite_func_fast() which uses the faster Cauchy integral
   algorithm in the same paper by Bunck

** add gsl_vector_axpby()

** add un-pivoted LDLT decomposition and its banded
   variant (gsl_linalg_ldlt_* and gsl_linalg_ldlt_band_*)

** add binary search tree module (gsl_bst); based on GNU libavl

** remove -u flag to gsl-histogram

** updated spmatrix module
   - added routines and data structures for all types (float,uint,char,...)
   - added gsl_spmatrix_scale_columns() and gsl_spmatrix_scale_rows()
   - added gsl_spmatrix_add_to_dense()
   - more efficient reallocation of COO/triplet matrices (no longer rebuilds binary tree)
   - enhanced test suite
   - added gsl_spmatrix_min_index()

** add routines for banded Cholesky decomposition (gsl_linalg_cholesky_band_*)

** documented gsl_linalg_LQ routines and added gsl_linalg_LQ_lssolve()

* What was new in gsl-2.5:

** doc bug fix in binomial distribution figure (Damien Desfontaines)

** added Wishart distribution (Timothée Flutre)

** added new module for digital filtering (gsl_filter); current filters include:
     Gaussian filter
     median filter
     recursive median filter
     impulse detection filter

** added new module for moving window statistics (gsl_movstat)

** added statistics functions:
     gsl_stats_median()
     gsl_stats_select()
     gsl_stats_mad()
     gsl_stats_mad0()
     gsl_stats_Sn_from_sorted_data()
     gsl_stats_Qn_from_sorted_data()
     gsl_stats_gastwirth_from_sorted_data()
     gsl_stats_trmean_from_sorted_data()

** added Romberg integration (gsl_integration_romberg)

** bug fix in deprecated functions gsl_multifit_wlinear_svd and
   gsl_multifit_wlinear_usvd (reported by Vlad Koli)

** documentation corrected to state that gsl_sf_legendre functions do

gsl-randist.c
/* randist/gsl-randist.c
 * 
 * Copyright (C) 1996, 1997, 1998, 1999, 2000 James Theiler, Brian Gough
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or (at
 * your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

#include <config.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>

#include <gsl/gsl_randist.h>
#include <gsl/gsl_rng.h>
#include <gsl/gsl_test.h>

void error (const char * s);


int
main (int argc, char *argv[])
{
  size_t i,j;
  size_t n = 0;
  double mu = 0, nu = 0, nu1 = 0, nu2 = 0, sigma = 0, a = 0, b = 0, c = 0;
  double zeta = 0, sigmax = 0, sigmay = 0, rho = 0;
  double p = 0;
  double x = 0, y =0, z=0  ;
  unsigned int N = 0, t = 0, n1 = 0, n2 = 0 ;
  unsigned long int seed = 0 ;
  const char * name ;
  gsl_rng * r ;

  if (argc < 4) 
    {
      printf (
"Usage: gsl-randist seed n DIST param1 param2 ...\n"
"Generates n samples from the distribution DIST with parameters param1,\n"
"param2, etc. Valid distributions are,\n"
"\n"
"  beta\n"
"  binomial\n"
"  bivariate-gaussian\n"
"  cauchy\n"
"  chisq\n"
"  dir-2d\n"
"  dir-3d\n"
"  dir-nd\n"
"  erlang\n"
"  exponential\n"
"  exppow\n"
"  fdist\n"
"  flat\n"
"  gamma\n"
"  gaussian-tail\n"
"  gaussian\n"
"  geometric\n"
"  gumbel1\n"
"  gumbel2\n"
"  hypergeometric\n"
"  laplace\n"
"  landau\n"
"  levy\n"
"  levy-skew\n"
"  logarithmic\n"
"  logistic\n"
"  lognormal\n"
"  negative-binomial\n"
"  pareto\n"
"  pascal\n"
"  poisson\n"
"  rayleigh-tail\n"
"  rayleigh\n"
"  tdist\n"
"  ugaussian-tail\n"
"  ugaussian\n"
"  weibull\n") ;
      exit (0);
    }

  argv++ ; seed = atol (argv[0]); argc-- ;
  argv++ ; n = atol (argv[0]); argc-- ;
  argv++ ; name = argv[0] ; argc-- ; argc-- ;

  gsl_rng_env_setup() ;

  if (gsl_rng_default_seed != 0) {
    fprintf(stderr, 
            "overriding GSL_RNG_SEED with command line value, seed = %ld\n", 
            seed) ;
  }
  
  gsl_rng_default_seed = seed ;

  r = gsl_rng_alloc(gsl_rng_default) ;


#define NAME(x) !strcmp(name,(x))
#define OUTPUT(x) for (i = 0; i < n; i++) { printf("%g\n", (x)) ; }
#define OUTPUT1(a,x) for(i = 0; i < n; i++) { a ; printf("%g\n", x) ; }
#define OUTPUT2(a,x,y) for(i = 0; i < n; i++) { a ; printf("%g %g\n", x, y) ; }
#define OUTPUT3(a,x,y,z) for(i = 0; i < n; i++) { a ; printf("%g %g %g\n", x, y, z) ; }
#define INT_OUTPUT(x) for (i = 0; i < n; i++) { printf("%d\n", (x)) ; }
#define ARGS(x,y) if (argc != x) error(y) ;
#define DBL_ARG(x) if (argc) { x=atof((++argv)[0]);argc--;} else {error( #x);};
#define INT_ARG(x) if (argc) { x=atoi((++argv)[0]);argc--;} else {error( #x);};

  if (NAME("bernoulli"))
    {
      ARGS(1, "p = probability of success");
      DBL_ARG(p)
      INT_OUTPUT(gsl_ran_bernoulli (r, p));
    }
  else if (NAME("beta"))
    {
      ARGS(2, "a,b = shape parameters");
      DBL_ARG(a)
      DBL_ARG(b)
      OUTPUT(gsl_ran_beta (r, a, b));
    }
  else if (NAME("binomial"))
    {
      ARGS(2, "p = probability, N = number of trials");
      DBL_ARG(p)
      INT_ARG(N)
      INT_OUTPUT(gsl_ran_binomial (r, p, N));
    }
  else if (NAME("cauchy"))
    {
      ARGS(1, "a = scale parameter");
      DBL_ARG(a)
      OUTPUT(gsl_ran_cauchy (r, a));
    }
  else if (NAME("chisq"))
    {
      ARGS(1, "nu = degrees of freedom");
      DBL_ARG(nu)
      OUTPUT(gsl_ran_chisq (r, nu));
    }
  else if (NAME("erlang"))
    {
      ARGS(2, "a = scale parameter, b = order");
      DBL_ARG(a)
      DBL_ARG(b)
      OUTPUT(gsl_ran_erlang (r, a, b));
    }
  else if (NAME("exponential"))
    {
      ARGS(1, "mu = mean value");
      DBL_ARG(mu) ;
      OUTPUT(gsl_ran_exponential (r, mu));
    }
  else if (NAME("exppow"))
    {
      ARGS(2, "a = scale parameter, b = power (1=exponential, 2=gaussian)");
      DBL_ARG(a) ;
      DBL_ARG(b) ;
      OUTPUT(gsl_ran_exppow (r, a, b));
    }
  else if (NAME("fdist"))
    {
      ARGS(2, "nu1, nu2 = degrees of freedom parameters");
      DBL_ARG(nu1) ;
      DBL_ARG(nu2) ;
      OUTPUT(gsl_ran_fdist (r, nu1, nu2));
    }
  else if (NAME("flat"))
    {
      ARGS(2, "a = lower limit, b = upper limit");
      DBL_ARG(a) ;
      DBL_ARG(b) ;
      OUTPUT(gsl_ran_flat (r, a, b));
    }
  else if (NAME("gamma"))
    {
      ARGS(2, "a = order, b = scale");
      DBL_ARG(a) ;
      DBL_ARG(b) ;
      OUTPUT(gsl_ran_gamma (r, a, b));
    }
  else if (NAME("gaussian"))
    {
      ARGS(1, "sigma = standard deviation");
      DBL_ARG(sigma) ;
      OUTPUT(gsl_ran_gaussian (r, sigma));
    }
  else if (NAME("gaussian-tail"))
    {
      ARGS(2, "a = lower limit, sigma = standard deviation");
      DBL_ARG(a) ;
      DBL_ARG(sigma) ;
      OUTPUT(gsl_ran_gaussian_tail (r, a, sigma));
    }
  else if (NAME("ugaussian"))
    {
      ARGS(0, "unit gaussian, no parameters required");
      OUTPUT(gsl_ran_ugaussian (r));
    }
  else if (NAME("ugaussian-tail"))
    {
      ARGS(1, "a = lower limit");
      DBL_ARG(a) ;
      OUTPUT(gsl_ran_ugaussian_tail (r, a));
    }
  else if (NAME("bivariate-gaussian"))
    {
      ARGS(3, "sigmax = x std.dev., sigmay = y std.dev., rho = correlation");
      DBL_ARG(sigmax) ;
      DBL_ARG(sigmay) ;
      DBL_ARG(rho) ;
      OUTPUT2(gsl_ran_bivariate_gaussian (r, sigmax, sigmay, rho, &x, &y), 
              x, y);
    }
  else if (NAME("dir-2d"))
    {
      OUTPUT2(gsl_ran_dir_2d (r, &x, &y), x, y);
    }
  else if (NAME("dir-3d"))
    {
      OUTPUT3(gsl_ran_dir_3d (r, &x, &y, &z), x, y, z);
    }
  else if (NAME("dir-nd"))
    {
      double *xarr;  
      ARGS(1, "n1 = number of dimensions of hypersphere"); 
      INT_ARG(n1) ;
      xarr = (double *)malloc(n1*sizeof(double));

      for(i = 0; i < n; i++) { 
        gsl_ran_dir_nd (r, n1, xarr) ; 
        for (j = 0; j < n1; j++) { 
          if (j) putchar(' '); 
          printf("%g", xarr[j]) ; 
        } 
        putchar('\n'); 
      } ;

      free(xarr);
    }  
  else if (NAME("geometric"))
    {
      ARGS(1, "p = bernoulli trial probability of success");
      DBL_ARG(p) ;
      INT_OUTPUT(gsl_ran_geometric (r, p));
    }
  else if (NAME("gumbel1"))
    {
      ARGS(2, "a = order, b = scale parameter");
      DBL_ARG(a) ;
      DBL_ARG(b) ;
      OUTPUT(gsl_ran_gumbel1 (r, a, b));
    }
  else if (NAME("gumbel2"))
    {
      ARGS(2, "a = order, b = scale parameter");
      DBL_ARG(a) ;
      DBL_ARG(b) ;
      OUTPUT(gsl_ran_gumbel2 (r, a, b));
    }
  else if (NAME("hypergeometric"))
    {
      ARGS(3, "n1 = tagged population, n2 = untagged population, t = number of trials");
      INT_ARG(n1) ;
      INT_ARG(n2) ;
      INT_ARG(t) ;
      INT_OUTPUT(gsl_ran_hypergeometric (r, n1, n2, t));
    }
  else if (NAME("laplace"))
    {
      ARGS(1, "a = scale parameter");
      DBL_ARG(a) ;
      OUTPUT(gsl_ran_laplace (r, a));
    }
  else if (NAME("landau"))
    {
      ARGS(0, "no arguments required");
      OUTPUT(gsl_ran_landau (r));
    }
  else if (NAME("levy"))
    {
      ARGS(2, "c = scale, a = power (1=cauchy, 2=gaussian)");
      DBL_ARG(c) ;
      DBL_ARG(a) ;
      OUTPUT(gsl_ran_levy (r, c, a));
    }
  else if (NAME("levy-skew"))
    {
      ARGS(3, "c = scale, a = power (1=cauchy, 2=gaussian), b = skew");
      DBL_ARG(c) ;
      DBL_ARG(a) ;
      DBL_ARG(b) ;
      OUTPUT(gsl_ran_levy_skew (r, c, a, b));
    }
  else if (NAME("logarithmic"))
    {
      ARGS(1, "p = probability");
      DBL_ARG(p) ;
      INT_OUTPUT(gsl_ran_logarithmic (r, p));
    }
  else if (NAME("logistic"))
    {
      ARGS(1, "a = scale parameter");
      DBL_ARG(a) ;
      OUTPUT(gsl_ran_logistic (r, a));
    }
  else if (NAME("lognormal"))
    {
      ARGS(2, "zeta = location parameter, sigma = scale parameter");
      DBL_ARG(zeta) ;
      DBL_ARG(sigma) ;
      OUTPUT(gsl_ran_lognormal (r, zeta, sigma));
    }
  else if (NAME("negative-binomial"))
    {
      ARGS(2, "p = probability, a = order");
      DBL_ARG(p) ;
      DBL_ARG(a) ;
      INT_OUTPUT(gsl_ran_negative_binomial (r, p, a));
    }
  else if (NAME("pareto"))
    {
      ARGS(2, "a = power, b = scale parameter");
      DBL_ARG(a) ;
      DBL_ARG(b) ;
      OUTPUT(gsl_ran_pareto (r, a, b));
    }
  else if (NAME("pascal"))
    {
      ARGS(2, "p = probability, n = order (integer)");
      DBL_ARG(p) ;
      INT_ARG(N) ;
      INT_OUTPUT(gsl_ran_pascal (r, p, N));
    }
  else if (NAME("poisson"))
    {
      ARGS(1, "mu = scale parameter");
      DBL_ARG(mu) ;
      INT_OUTPUT(gsl_ran_poisson (r, mu));
    }
  else if (NAME("rayleigh"))
    {
      ARGS(1, "sigma = scale parameter");
      DBL_ARG(sigma) ;
      OUTPUT(gsl_ran_rayleigh (r, sigma));
    }
  else if (NAME("rayleigh-tail"))
    {
      ARGS(2, "a = lower limit, sigma = scale parameter");
      DBL_ARG(a) ;
      DBL_ARG(sigma) ;
      OUTPUT(gsl_ran_rayleigh_tail (r, a, sigma));
    }
  else if (NAME("tdist"))
    {
      ARGS(1, "nu = degrees of freedom");
      DBL_ARG(nu) ;
      OUTPUT(gsl_ran_tdist (r, nu));
    }
  else if (NAME("weibull"))
    {
      ARGS(2, "a = scale parameter, b = exponent");
      DBL_ARG(a) ;
      DBL_ARG(b) ;
      OUTPUT(gsl_ran_weibull (r, a, b));
    }
  else
    {
      fprintf(stderr,"Error: unrecognized distribution: %s\n", name) ;
    }

  return 0 ;
}


void
error (const char * s)
{
  fprintf(stderr, "Error: arguments should be %s\n",s) ;
  exit (EXIT_FAILURE) ;
}
